static_frame/core/index_hierarchy.py

```diff
@@ -605,62 +862,69 @@ def copy(self: IH) -> IH:
     # name interface
 
-    def rename(self: IH, name: NameType) -> IH:
+    def rename(self: IH,
+            name: NameType,
+            ) -> IH:
         '''
-        Return a new Frame with an updated name attribute.
+        Return a new IndexHierarchy with an updated name attribute.
         '''
-        # do not need to recache
-        # let the constructor handle reuse
-        return self.__class__(self, name=name)
+        if self.STATIC:
+            indices = self._indices
+            blocks = self._blocks
+        else:
+            indices = [idx.copy() for idx in self._indices]
+            blocks = self._blocks.copy()
+
+        return self.__class__(
+                indices=indices,
+                indexers=list(self._indexers),
+                name=name,
+                blocks=blocks,
+                own_blocks=True,
+                )
 
-    #---------------------------------------------------------------------------
+    # --------------------------------------------------------------------------
     # interfaces
 
     @property
-    def loc(self) -> InterfaceGetItem['IndexHierarchy']:
-        return InterfaceGetItem(self._extract_loc) #type: ignore
+    def loc(self: IH) -> InterfaceGetItem['IndexHierarchy']:
+        return InterfaceGetItem(self._extract_loc)
 
     @property
-    def iloc(self) -> InterfaceGetItem['IndexHierarchy']:
-        return InterfaceGetItem(self._extract_iloc) #type: ignore
-
+    def iloc(self: IH) -> InterfaceGetItem['IndexHierarchy']:
+        return InterfaceGetItem(self._extract_iloc)
 
-    def _iter_label(self,
+    def _iter_label(self: IH,
             depth_level: tp.Optional[DepthLevelSpecifier] = None,
             ) -> tp.Iterator[tp.Hashable]:
+        '''
+        Iterate over labels at a given depth level.
 
+        For multiple depth levels, the iterator will yield tuples of labels.
+        '''
         if depth_level is None: # default to full labels
             depth_level = list(range(self.depth))
 
-        # if no type blocks, use a levels
-        if self._recache:
-            if isinstance(depth_level, INT_TYPES):
-                yield from self._levels.labels_at_depth(depth_level=depth_level)
-            else:
-                yield from zip(
-                        *(self._levels.labels_at_depth(depth_level=d) for d in depth_level)
-                        )
+        if isinstance(depth_level, INT_TYPES):
+            yield from self.values_at_depth(depth_level)
         else:
-            if isinstance(depth_level, INT_TYPES):
-                yield from self._blocks._extract_array(column_key=depth_level)
-            else:
-                yield from array2d_to_tuples(
-                        self._blocks._extract_array(column_key=depth_level)
-                        )
+            yield from array2d_to_tuples(
+                    self._blocks._extract_array(column_key=depth_level)
+                    )
```

Comment:
Using `_extract_array` here may force undesirable type coercion; better to `zip` calls to `self.values_at_depth`, or find an alternative way to build these tuples.